### ISSUE TITLE
`.clang_arg("-x c++")` causes failure in `generate`

### ISSUE THREAD

**Issue by kulp** [rust-lang/rust-bindgen#1855]
As reported by @virtualritz in #1852, using `clang_arg("-x c++")` causes `bindgen`'s `generate` to return an `Err` result.

From this `build.rs`,
```rust
use std::path::PathBuf;

fn main() {
    let bindings = bindgen::Builder::default()
        .header("wrapper.hpp")
        .clang_arg("-x c++")            // does not work
        .generate()
        .expect("Couldn't make bindings!");

    let out_path = PathBuf::from(std::env::var("OUT_DIR").unwrap());
    bindings.write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

`RUST_BACKTRACE=1 cargo build -vv` shows :
```
error: language not recognized: ' c++', err: true
warning: '-x c++': 'linker' input unused [-Wunused-command-line-argument], err: false
thread 'main' panicked at 'Couldn't make bindings!: ()', build.rs:4:20
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: std::panicking::rust_panic_with_hook
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: core::result::Result<T,E>::expect
  10: build_script_build::main
  11: std::rt::lang_start::{{closure}}
  12: std::rt::lang_start_internal
  13: std::rt::lang_start
  14: main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by kulp** [comment]
There is more explanation in https://github.com/rust-lang/rust-bindgen/pull/1852#issuecomment-667215433.

I think that this is basically a quoting issue. When a user invokes `clang` on the command line, `clang -x c++ wrapper.h` shows up in `argv` as :
```c
["clang", "-x", "c++", "wrapper.h"]
```
because the shell splits arguments on whitespace.

However, when a user of `bindgen` uses `.clang_arg("-x c++")`, no shell is involved, so no splitting on whitespace occurs. Therefore, `libclang` sees an array like this instead :
```c
["clang", "-x c++", "wrapper.h"]
```
and `libclang` interprets the language as >>>` c++`<<< **with a leading space**.

This is something that `bindgen` can detect and fix in code, without updating documentation.

**Comment by virtualritz** [comment]
I think the stance on my two PRs is not helpful until this is fixed.

Every time a developer runs into this they waste time tracking this down. It took me 15mins to find someone on Discord who understood what was going on and suggested this. Without that I would have been stuck. Other people may be less lucky.

What I am asking myself: why not change the docs for now and roll back the changes when this is fixed?

No need to answer. I consider my time a tad too valuable to exchange half a dozen messages over a PR to remove two spaces that should not be in the docs *as of this writing*. And which could easily be added again once they should.

**Comment by kulp** [comment]
@virtualritz, if we change the docs, should we also change all the other places where `-x c++` is mentioned in the headers ? If not, why not ?

I recognize that you were trying to be helpful by providing a solution rather than merely filing an issue.

However, sometimes that turns out to be the wrong decision, in my view.

If @emilio thinks differently, he can of course override me.

**Comment by kulp** [comment]
The good news is, thanks to your efforts, the next users who run into this will be able to find an issue filed about it -- this one. I have updated the description to show the workaround.

**Comment by virtualritz** [comment]
Yeah, sorry. Just had my grumpy five minutes. Espresso fixed it. Bindgen is amazing. 90% of the work I do with Rust for now depends on it. So I'm glad I could give back the tiniest bit.

Someone on Discord also mentioned the workaround of just using the `.hpp` extension for the wrapper header to force `bindgen` into C++ mode.
That may be the best solution for the docs for now as it will work without changing code before and after the fix – it doesn't require specifying the `-x` flag at all.

**Comment by kulp** [comment]
> Someone on Discord also mentioned the workaround of just using the `.hpp` extension for the wrapper header to force `bindgen` into C++ mode.
> That may be the best solution for the docs for now as it will work without requiring crate users to change code before and after the fix – it doesn't require specifying the `-x` flag at all.

@virtualritz, that is not a bad idea. Although I myself do not have the authority to merge pull requests, I am cautiously optimistic that, if you felt inclined to add a sentence or two explaining that better than the docs currently do, that @emilio would authorize it.

Otherwise, I will try to see in the next week or so if I can get a permanent solution into the code.

**Comment by virtualritz** [comment]
> @virtualritz, that is not a bad idea. Although I myself do not have the authority to merge pull requests, I am cautiously optimistic that, if you felt inclined to add a sentence or two explaining that better than the docs currently do, that @emilio would authorize it.
@kulp Which parts of the docs would need this treatment? If @emilio agrees this makes sense I'm happy to write something up that can be kept even after this is fixed.

**Comment by kulp** [comment]
@virtualritz, I think it might be helpful to update :
- `book/src/command-line-usage.md`
    - give a C++ example with an `.hpp` file
    - give a C++ example with `-x c++`
        - This should work on the command line even with the space between `-x` and `c++`. Please test it if you can -- `cargo run` inside `rust-bindgen` will run the bindgen command-line tool.
- `book/src/cpp.md`
    - give an explicit example of a call to `.clang_args(&["-x", "c++"])`, which does work currently
- `book/src/objc.md`
    - show a similar example as in `cpp.md`

These are just my ideas. I do not see any other places that would need to be updated. See what you think.

**Comment by kulp** [comment]
At some point (perhaps now, perhaps not), we could consider a larger change that prefers [`--language=c++`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-x-language) over `-x c++`, avoiding the problem of the space, and also making the argument more self-documenting. I have not checked whether `--language=xyz` is supported back in `libclang` 3.9, though -- we need, for now, to remain compatible with that old `libclang` for a little while yet.

**Comment by emilio** [comment]
Yes, we really need to document this clearly. We could make `clang_arg` more auto-magical I guess, but that breaks paths with spaces and so on, so not a fan.

@kulp's suggestions are great. Suggesting `--language=c++` is probably also ok if libclang supports it (it probably does).

**Comment by kulp** [comment]
> We could make `clang_arg` more auto-magical I guess, but that breaks paths with spaces and so on, so not a fan.

@emilio, I definitely agree that making `.clang_arg()` automagical regarding whitespace is not prudent. However, since we already check for various forms of `-xc++` :

https://github.com/rust-lang/rust-bindgen/blob/b424c17608206166b98db0e1b7da752a1e560f54/src/lib.rs#L97-L101

then perhaps `generate()` could narrowly change `-x lang` specifically, either into `-xlang` or into `--language=lang`. It would still be a hack, but it would solve this particular problem, and it could make `args_are_cpp` simpler. Would that be worthwhile ?

If that is not a great idea then all I can think of is to improve the docs and examples, which would be valuable in any case.

**Comment by yo123abxd** [comment]
> 

It works well !!!
